stdlib/public/runtime/ImageInspectionCommon.h

```diff
@@ -57,14 +57,27 @@ struct SectionInfo {
   const char *data;
 };
 
-// Called by injected constructors when a dynamic library is loaded.
+/// Called by injected constructors when a dynamic library is loaded.
+///
+/// \param sections A structure describing the metadata sections in the
+///     newly-loaded image.
+///
+/// \warning The runtime keeps a reference to \a sections and may mutate it, so
+///   it \em must be mutable and long-lived (that is, statically or dynamically
+///   allocated.) The effect of passing a pointer to a local value is undefined.
 SWIFT_RUNTIME_EXPORT
-void swift_addNewDSOImage(const void *addr);
+void swift_addNewDSOImage(struct swift::MetadataSections *sections);
```

Comment:
The reason to keep the parameter as `void *` was to keep the header compatible across C/C++.  The registrar itself is meant to be in C, not C++, and so cannot process the `swift::MetadataSections *`.

Reply:
The registrars are written in C++ so this should not be an issue.